pkg/kubectl/rolling_updater.go

```diff
@@ -93,27 +79,436 @@ const (
 }
 
-// AlwaysAccept is an UpdateAcceptor which always accepts the controller.
-type AlwaysAccept struct{}
+// scaleUp scales up newRc to desired by increment. It will safely no-op as
+// necessary when it detects redundancy or other relevant conditions.
+func (r *RollingUpdater) scaleUp(newRc, oldRc *api.ReplicationController, desired, increment int, config *RollingUpdaterConfig) (*api.ReplicationController, error) {
+	// If we're already at the desired, do nothing.
+	if newRc.Spec.Replicas == desired {
+		return newRc, nil
+	}
+	// If the current safe increment is 0, we don't have capacity to scale up
+	// while maintaining in-place semantics, so do nothing.
+	if increment == 0 {
+		return newRc, nil
+	}
+	// If the old is already scaled down, go ahead and scale all the way up.
+	if oldRc.Spec.Replicas == 0 {
+		increment = desired - newRc.Spec.Replicas
+	}
+	// Increase the replica count, and deal with fenceposts.
+	newRc.Spec.Replicas += increment
+	if newRc.Spec.Replicas > desired {
+		newRc.Spec.Replicas = desired
+	}
+	// Perform the scale-up.
+	fmt.Fprintf(config.Out, "Scaling %s up to %d\n", newRc.Name, newRc.Spec.Replicas)
+	retryWait := &RetryParams{config.Interval, config.Timeout}
+	scaledRc, err := r.scaleAndWait(newRc, retryWait, retryWait)
+	if err != nil {
+		return nil, err
+	}
+	return scaledRc, nil
+}
+
+// scaleDown scales down oldRc to 0 by increment. It will safely no-op as
+// necessary when it detects redundancy or other relevant conditions.
+func (r *RollingUpdater) scaleDown(newRc, oldRc *api.ReplicationController, desired, maxIncrement, minReady int, checkType availabilityCheckType, config *RollingUpdaterConfig) (*api.ReplicationController, int, error) {
+	// Already scaled down; do nothing.
+	if oldRc.Spec.Replicas == 0 {
+		return oldRc, maxIncrement, nil
+	}
+	// Block until there are any pods available.
+	waitForAvailablePods := r.availabilityChecks[checkType]
+	_, newReady, ready, err := waitForAvailablePods(config.Interval, config.Timeout, oldRc, newRc)
+	if err != nil {
+		return nil, 0, err
+	}
+	// The increment will be as much as possible; if the update was started in a
+	// down-up configuration, we might only be able to scale down partially.
+	increment := int(math.Min(float64(maxIncrement), float64(ready-minReady)))
+	// The increment normally shouldn't drop below 0 because the ready count
+	// always starts below the old replica count, but the old replica count can
+	// decrement due to externalities like pods death in the replica set. This
+	// will be considered a transient condition; do nothing and try again later
+	// with new readiness values.
+	if increment < 0 {
+		return oldRc, 0, nil
+	}
+	// If the most we can scale is 0, it means we can't scale down without
+	// violating the minimum. Do nothing and try again later when there are more
+	// new ready (or more old ready if we're in a recovery from a negative
+	// increment.)
+	if increment == 0 {
+		return oldRc, increment, nil
```

Comment:
Returning `0` seems clearer, yeah. I decided to keep separate `if` blocks because each condition had so much explanation and I didn't want to make it even more confusing. :grin: